File: src/main/java/com/company/plugin/completion/ZyCompletionContributor.java
package com.company.plugin.completion;

import com.intellij.codeInsight.completion.*;
import com.intellij.codeInsight.lookup.LookupElement;
import com.intellij.codeInsight.lookup.LookupElementBuilder;
import com.intellij.openapi.diagnostic.Logger;
import com.intellij.openapi.vfs.VirtualFile;
import com.intellij.patterns.PlatformPatterns;
import com.intellij.psi.PsiElement;
import com.intellij.util.ProcessingContext;
import com.intellij.openapi.project.Project;
import org.jetbrains.annotations.NotNull;

import java.util.*;

/**
 * ZY 代码补全贡献者
 * 基于本地 Java 实现提供代码补全功能，不依赖 LSP
 */
public class ZyCompletionContributor extends CompletionContributor {
    
    private static final Logger LOG = Logger.getInstance(ZyCompletionContributor.class);
    
    // ZY 语言关键字
    private static final Set<String> ZY_KEYWORDS = new HashSet<String>() {{
        add("function");
        add("if");
        add("else");
        add("for");
        add("while");
        add("return");
        add("var");
        add("let");
        add("const");
        add("true");
        add("false");
        add("null");
        add("undefined");
        add("import");
        add("export");
        add("class");
        add("struct");
        add("interface");
        add("extends");
        add("implements");
        add("public");
        add("private");
        add("protected");
        add("static");
        add("abstract");
        add("final");
    }};
    
    // PHP 关键字
    private static final Set<String> PHP_KEYWORDS = new HashSet<String>() {{
        add("if");
        add("else");
        add("elseif");
        add("endif");
        add("while");
        add("endwhile");
        add("for");
        add("foreach");
        add("endforeach");
        add("do");
        add("switch");
        add("case");
        add("default");
        add("break");
        add("continue");
        add("goto");
        add("function");
        add("class");
        add("struct");
        add("interface");
        add("trait");
        add("namespace");
        add("use");
        add("as");
        add("public");
        add("private");
        add("protected");
        add("static");
        add("abstract");
        add("final");
        add("extends");
        add("implements");
        add("new");
        add("clone");
        add("instanceof");
        add("global");
        add("const");
        add("define");
        add("defined");
        add("try");
        add("catch");
        add("finally");
        add("throw");
        add("include");
        add("include_once");
        add("require");
        add("require_once");
        add("return");
        add("yield");
        add("echo");
        add("print");
        add("print_r");
        add("var_dump");
        add("isset");
        add("empty");
        add("unset");
        add("die");
        add("exit");
        add("array");
        add("string");
        add("int");
        add("float");
        add("bool");
        add("object");
        add("mixed");
        add("void");
        add("callable");
        add("iterable");
        add("resource");
        add("null");
        add("false");
        add("true");
    }};
    
    // PHP 内置函数
    private static final Set<String> PHP_FUNCTIONS = new HashSet<String>() {{
        // 字符串函数
        add("strlen");
        add("strpos");
        add("str_replace");
        add("substr");
        add("trim");
        add("ltrim");
        add("rtrim");
        add("strtolower");
        add("strtoupper");
        add("ucfirst");
        add("ucwords");
        add("str_split");
        add("explode");
        add("implode");
        add("sprintf");
        add("printf");
        add("htmlspecialchars");
        add("htmlentities");
        add("strip_tags");
        
        // 数组函数
        add("array");
        add("count");
        add("sizeof");
        add("array_push");
        add("array_pop");
        add("array_shift");
        add("array_unshift");
        add("array_merge");
        add("array_slice");
        add("array_splice");
        add("array_keys");
        add("array_values");
        add("array_search");
        add("in_array");
        add("array_unique");
        add("array_reverse");
        add("sort");
        add("rsort");
        add("asort");
        add("arsort");
        add("ksort");
        add("krsort");
        add("usort");
        add("uasort");
        add("uksort");
        
        // 文件系统函数
        add("file_exists");
        add("is_file");
        add("is_dir");
        add("file_get_contents");
        add("file_put_contents");
        add("fopen");
        add("fclose");
        add("fread");
        add("fwrite");
        add("fgets");
        add("fgetc");
        add("feof");
        add("fseek");
        add("ftell");
        add("copy");
        add("move_uploaded_file");
        add("unlink");
        add("mkdir");
        add("rmdir");
        add("chmod");
        add("chown");
        
        // 数学函数
        add("abs");
        add("ceil");
        add("floor");
        add("round");
        add("min");
        add("max");
        add("rand");
        add("mt_rand");
        add("pi");
        add("pow");
        add("sqrt");
        add("sin");
        add("cos");
        add("tan");
        add("asin");
        add("acos");
        add("atan");
        add("log");
        add("exp");
        
        // 日期时间函数
        add("date");
        add("time");
        add("strtotime");
        add("mktime");
        add("getdate");
        add("gmdate");
        add("microtime");
        
        // 其他常用函数
        add("json_encode");
        add("json_decode");
        add("serialize");
        add("unserialize");
        add("var_dump");
        add("print_r");
        add("gettype");
        add("is_array");
        add("is_string");
        add("is_numeric");
        add("is_null");
        add("is_bool");
        add("filter_var");
        add("filter_input");
        add("preg_match");
        add("preg_replace");
        add("preg_split");
    }};
    
    // PHP 常量
    private static final Set<String> PHP_CONSTANTS = new HashSet<String>() {{
        add("PHP_VERSION");
        add("PHP_OS");
        add("PHP_EOL");
        add("PHP_INT_MAX");
        add("PHP_INT_MIN");
        add("PHP_FLOAT_MAX");
        add("PHP_FLOAT_MIN");
        add("PHP_SAPI");
        add("PHP_BINARY");
        add("E_ERROR");
        add("E_WARNING");
        add("E_PARSE");
        add("E_NOTICE");
        add("E_CORE_ERROR");
        add("E_CORE_WARNING");
        add("E_COMPILE_ERROR");
        add("E_COMPILE_WARNING");
        add("E_USER_ERROR");
        add("E_USER_WARNING");
        add("E_USER_NOTICE");
        add("E_STRICT");
        add("E_RECOVERABLE_ERROR");
        add("E_DEPRECATED");
        add("E_USER_DEPRECATED");
        add("E_ALL");
        add("DIRECTORY_SEPARATOR");
        add("PATH_SEPARATOR");
        add("FILE_SEPARATOR");
        add("TRUE");
        add("FALSE");
        add("NULL");
        add("M_PI");
        add("M_E");
        add("M_LOG2E");
        add("M_LOG10E");
        add("M_LN2");
        add("M_LN10");
    }};
    
    public ZyCompletionContributor() {
        extend(
            CompletionType.BASIC,
            PlatformPatterns.psiElement(),
            new ZyCompletionProvider()
        );
    }
    
    private static class ZyCompletionProvider extends CompletionProvider<CompletionParameters> {
        
        @Override
        protected void addCompletions(
                @NotNull CompletionParameters parameters,
                @NotNull ProcessingContext context,
                @NotNull CompletionResultSet result
        ) {
            Project project = parameters.getEditor().getProject();
            if (project == null) return;
            
            VirtualFile file = parameters.getOriginalFile().getVirtualFile();
            if (file == null) return;
            
            // 只处理 .zy 文件
            if (!file.getName().endsWith(".zy")) {
                return;
            }
            
            try {
                LOG.debug("Adding completions for file: " + file.getName());
                
                // 获取当前光标位置的文本上下文
                String prefix = getPrefix(parameters);
                
                // 添加关键字补全
                addKeywordCompletions(result, prefix);
                
                // 添加 PHP 关键字补全
                addPhpKeywordCompletions(result, prefix);
                
                // 添加 PHP 函数补全
                addPhpFunctionCompletions(result, prefix);
                
                // 添加 PHP 常量补全
                addPhpConstantCompletions(result, prefix);
                
                // 添加代码片段补全
                addSnippetCompletions(result, prefix);
                
                LOG.debug("Added completions with prefix: " + prefix);
                
            } catch (Exception e) {
                LOG.error("Error in completion", e);
            }
        }
        
        /**
         * 获取当前光标位置的前缀文本
         */
        private String getPrefix(CompletionParameters parameters) {
            var document = parameters.getEditor().getDocument();
            int offset = parameters.getOffset();
            
            // 向前查找，直到遇到空白字符或特殊字符
            int start = offset;
            while (start > 0) {
                char c = document.getCharsSequence().charAt(start - 1);
                if (Character.isWhitespace(c) || !Character.isLetterOrDigit(c) && c != '_' && c != '$') {
                    break;
                }
                start--;
            }
            
            return document.getCharsSequence().subSequence(start, offset).toString();
        }
        
        /**
         * 添加 ZY 关键字补全
         */
        private void addKeywordCompletions(CompletionResultSet result, String prefix) {
            for (String keyword : ZY_KEYWORDS) {
                if (keyword.toLowerCase().startsWith(prefix.toLowerCase())) {
                    result.addElement(
                        LookupElementBuilder.create(keyword)
                            .withTypeText("ZY Keyword")
                            .withIcon(com.intellij.icons.AllIcons.Nodes.Static)
                    );
                }
            }
        }
        
        /**
         * 添加 PHP 关键字补全
         */
        private void addPhpKeywordCompletions(CompletionResultSet result, String prefix) {
            for (String keyword : PHP_KEYWORDS) {
                if (keyword.toLowerCase().startsWith(prefix.toLowerCase())) {
                    result.addElement(
                        LookupElementBuilder.create(keyword)
                            .withTypeText("PHP Keyword")
                            .withIcon(com.intellij.icons.AllIcons.Nodes.Static)
                    );
                }
            }
        }
        
        /**
         * 添加 PHP 函数补全
         */
        private void addPhpFunctionCompletions(CompletionResultSet result, String prefix) {
            for (String function : PHP_FUNCTIONS) {
                if (function.toLowerCase().startsWith(prefix.toLowerCase())) {
                    result.addElement(
                        LookupElementBuilder.create(function + "()")
                            .withTypeText("PHP Function")
                            .withIcon(com.intellij.icons.AllIcons.Nodes.Function)
                            .withInsertHandler((context, item) -> {
                                // 将光标定位到括号中间
                                int offset = context.getStartOffset() + function.length() + 1;
                                context.getEditor().getCaretModel().moveToOffset(offset);
                            })
                    );
                }
            }
        }
        
        /**
         * 添加 PHP 常量补全
         */
        private void addPhpConstantCompletions(CompletionResultSet result, String prefix) {
            for (String constant : PHP_CONSTANTS) {
                if (constant.toLowerCase().startsWith(prefix.toLowerCase())) {
                    result.addElement(
                        LookupElementBuilder.create(constant)
                            .withTypeText("PHP Constant")
                            .withIcon(com.intellij.icons.AllIcons.Nodes.Constant)
                    );
                }
            }
        }
        
        /**
         * 添加代码片段补全
         */
        private void addSnippetCompletions(CompletionResultSet result, String prefix) {
            // 函数定义片段
            if ("function".startsWith(prefix.toLowerCase())) {
                result.addElement(
                    LookupElementBuilder.create("function")
                        .withTypeText("Function Definition")
                        .withIcon(com.intellij.icons.AllIcons.Nodes.Function)
                        .withInsertHandler((context, item) -> {
                            // 先删除已经插入的 "function" 文本
                            int startOffset = context.getStartOffset();
                            int endOffset = startOffset + "function".length();
                            context.getDocument().deleteString(startOffset, endOffset);
                            
                            // 插入完整的函数定义代码片段
                            String snippet = "function ${1:functionName}(${2:parameters}) {\n    ${3:// function body}\n    return ${4:value};\n}";
                            insertSnippet(context, snippet);
                        })
                );
            }
            
            // if 语句片段
            if ("if".startsWith(prefix.toLowerCase())) {
                result.addElement(
                    LookupElementBuilder.create("if")
                        .withTypeText("If Statement")
                            .withIcon(com.intellij.icons.AllIcons.Nodes.Static)
                        .withInsertHandler((context, item) -> {
                            // 先删除已经插入的 "if" 文本
                            int startOffset = context.getStartOffset();
                            int endOffset = startOffset + "if".length();
                            context.getDocument().deleteString(startOffset, endOffset);
                            
                            // 插入完整的 if 语句代码片段
                            String snippet = "if (${1:condition}) {\n    ${2:// code}\n}";
                            insertSnippet(context, snippet);
                        })
                );
            }
            
            // for 循环片段
            if ("for".startsWith(prefix.toLowerCase())) {
                result.addElement(
                    LookupElementBuilder.create("for")
                        .withTypeText("For Loop")
                            .withIcon(com.intellij.icons.AllIcons.Nodes.Static)
                        .withInsertHandler((context, item) -> {
                            // 先删除已经插入的 "for" 文本
                            int startOffset = context.getStartOffset();
                            int endOffset = startOffset + "for".length();
                            context.getDocument().deleteString(startOffset, endOffset);
                            
                            // 插入完整的 for 循环代码片段
                            String snippet = "for (${1:initialization}; ${2:condition}; ${3:increment}) {\n    ${4:// code}\n}";
                            insertSnippet(context, snippet);
                        })
                );
            }
            
            // while 循环片段
            if ("while".startsWith(prefix.toLowerCase())) {
                result.addElement(
                    LookupElementBuilder.create("while")
                        .withTypeText("While Loop")
                            .withIcon(com.intellij.icons.AllIcons.Nodes.Static)
                        .withInsertHandler((context, item) -> {
                            // 先删除已经插入的 "while" 文本
                            int startOffset = context.getStartOffset();
                            int endOffset = startOffset + "while".length();
                            context.getDocument().deleteString(startOffset, endOffset);
                            
                            // 插入完整的 while 循环代码片段
                            String snippet = "while (${1:condition}) {\n    ${2:// code}\n}";
                            insertSnippet(context, snippet);
                        })
                );
            }
            
            // class 定义片段
            if ("class".startsWith(prefix.toLowerCase())) {
                result.addElement(
                    LookupElementBuilder.create("class")
                        .withTypeText("Class Definition")
                        .withIcon(com.intellij.icons.AllIcons.Nodes.Class)
                        .withInsertHandler((context, item) -> {
                            // 先删除已经插入的 "class" 文本
                            int startOffset = context.getStartOffset();
                            int endOffset = startOffset + "class".length();
                            context.getDocument().deleteString(startOffset, endOffset);
                            
                            // 插入完整的类定义代码片段
                            String snippet = "class ${1:ClassName} {\n    ${2:// class body}\n}";
                            insertSnippet(context, snippet);
                        })
                );
            }
        }
        
        /**
         * 插入代码片段
         */
        private void insertSnippet(InsertionContext context, String snippet) {
            // 处理代码片段，将 ${n:placeholder} 替换为占位符文本
            String processedSnippet = snippet.replaceAll("\\$\\{([0-9]+):([^}]*)\\}", "$2");
            // 移除剩余的 ${n} 占位符（没有默认值的）
            processedSnippet = processedSnippet.replaceAll("\\$\\{([0-9]+)\\}", "");
            
            // 插入文本
            context.getDocument().insertString(context.getStartOffset(), processedSnippet);
            
            // 查找第一个占位符并定位光标
            String[] placeholders = {"functionName", "ClassName", "condition", "initialization", "parameters"};
            for (String placeholder : placeholders) {
                int placeholderIndex = processedSnippet.indexOf(placeholder);
                if (placeholderIndex >= 0) {
                    int offset = context.getStartOffset() + placeholderIndex;
                    context.getEditor().getCaretModel().moveToOffset(offset);
                    // 选中占位符文本，方便用户直接输入
                    context.getEditor().getSelectionModel().setSelection(offset, offset + placeholder.length());
                    break;
                }
            }
        }
    }
}

File: src/main/kotlin/com/company/plugin/completion/ZyCompletionContributor.kt
package com.company.plugin.completion

import com.company.plugin.lsp.ZyLspService
import com.intellij.codeInsight.completion.*
import com.intellij.codeInsight.lookup.LookupElement
import com.intellij.codeInsight.lookup.LookupElementBuilder
import com.intellij.openapi.diagnostic.Logger
import com.intellij.openapi.vfs.VirtualFile
import com.intellij.patterns.PlatformPatterns
import com.intellij.psi.PsiElement
import com.intellij.util.ProcessingContext
import org.eclipse.lsp4j.CompletionItem
import org.eclipse.lsp4j.CompletionItemKind
import org.eclipse.lsp4j.CompletionList
import org.eclipse.lsp4j.InsertTextFormat
import org.eclipse.lsp4j.TextEdit
import org.eclipse.lsp4j.Range
import org.eclipse.lsp4j.Position

/**
 * ZY 代码补全贡献者
 * 基于 LSP 提供代码补全功能
 */
class ZyCompletionContributor : CompletionContributor() {
    
    companion object {
        private val LOG = Logger.getInstance(ZyCompletionContributor::class.java)
    }
    
    init {
        extend(
            CompletionType.BASIC,
            PlatformPatterns.psiElement(),
            ZyLspCompletionProvider()
        )
    }
    
    private class ZyLspCompletionProvider : CompletionProvider<CompletionParameters>() {
        
        override fun addCompletions(
            parameters: CompletionParameters,
            context: ProcessingContext,
            result: CompletionResultSet
        ) {
            val project = parameters.editor.project ?: return
            val file = parameters.originalFile.virtualFile ?: return
            
            // 只处理 .zy 文件
            if (!file.name.endsWith(".zy")) {
                return
            }
            
            try {
                val lspService = project.getService(ZyLspService::class.java)
                if (lspService == null || !lspService.isStarted()) {
                    LOG.debug("LSP service not available or not started")
                    return
                }
                
                val document = parameters.editor.document
                val offset = parameters.offset
                val lineNumber = document.getLineNumber(offset)
                val lineStartOffset = document.getLineStartOffset(lineNumber)
                val character = offset - lineStartOffset
                
                val uri = file.url
                // 将当前文档内容 didOpen 到 LSP（只执行一次）
                val text = document.text
                lspService.ensureDidOpen(uri, "zy", text)
                
                // 通知 LSP 文档变更（确保同步最新内容）
                lspService.notifyDocumentChange(uri, text)
                
                LOG.debug("Requesting completion at ${file.name}:${lineNumber}:${character}")
                
                val completionFuture = lspService.getCompletion(uri, lineNumber, character)
                
                // 等待 LSP 响应（设置超时）
                try {
                    val completionResult = completionFuture.get(3, java.util.concurrent.TimeUnit.SECONDS)
                    
                    if (completionResult.isLeft) {
                        // List<CompletionItem>
                        val items = completionResult.left
                        for (item in items) {
                            result.addElement(createLookupElement(item, document, offset))
                        }
                        LOG.debug("Added ${items.size} completion items")
                    } else {
                        // CompletionList
                        val completionList = completionResult.right
                        for (item in completionList.items) {
                            result.addElement(createLookupElement(item, document, offset))
                        }
                        LOG.debug("Added ${completionList.items.size} completion items from list")
                    }
                } catch (e: Exception) {
                    LOG.debug("LSP completion request failed or timed out", e)
                }
                
            } catch (e: Exception) {
                LOG.error("Error in LSP completion", e)
            }
        }
        
        /**
         * 将 LSP CompletionItem 转换为 IntelliJ LookupElement
         */
        private fun createLookupElement(item: CompletionItem, document: com.intellij.openapi.editor.Document, offset: Int): LookupElement {
            // 使用 insertText 或 label 作为补全文本
            val lookupString = item.insertText ?: item.label
            
            var builder = LookupElementBuilder.create(lookupString)
                .withPresentableText(item.label)
            
            // 添加类型信息
            item.detail?.let { detail ->
                builder = builder.withTypeText(detail, true)
            }
            
            // 添加文档信息作为尾部文本
            item.documentation?.let { doc ->
                when {
                    doc.isLeft && doc.left is String -> {
                        val docString = doc.left as String
                        if (docString.isNotEmpty()) {
                            builder = builder.withTailText(" $docString", true)
                        }
                    }
                    doc.isRight -> {
                        // 处理 MarkupContent 类型的文档
                        val markupContent = doc.right
                        if (markupContent != null && markupContent.value.isNotEmpty()) {
                            val preview = if (markupContent.value.length > 50) {
                                markupContent.value.substring(0, 47) + "..."
                            } else {
                                markupContent.value
                            }
                            builder = builder.withTailText(" $preview", true)
                        }
                    }
                }
            }
            
            // 设置图标（根据 CompletionItemKind）
            val icon = getIconForKind(item.kind)
            if (icon != null) {
                builder = builder.withIcon(icon)
            }
            
            // 处理文本替换范围与 Snippet 扩展
            val textEdit = item.textEdit
            val isSnippet = item.insertTextFormat == InsertTextFormat.Snippet

            if (isSnippet) {
                // 代码片段：将 LSP Snippet 转换为可插入文本，并尽量定位到首个光标位
                val snippetText = item.insertText ?: when {
                    textEdit != null && textEdit.isLeft -> textEdit.left.newText
                    else -> item.label
                }
                val parsed = parseLspSnippet(snippetText)
                builder = builder.withInsertHandler { context, _ ->
                    val doc = context.document
                    val caretModel = context.editor.caretModel

                    // 先移除默认插入的 lookupString，避免重复文本
                    if (context.startOffset <= context.tailOffset) {
                        doc.replaceString(context.startOffset, context.tailOffset, "")
                    }

                    if (textEdit != null && textEdit.isLeft) {
                        val edit = textEdit.left
                        val startOffset = toOffset(doc, edit.range.start)
                        val endOffset = toOffset(doc, edit.range.end)
                        doc.replaceString(startOffset, endOffset, parsed.text)
                        val caretOffset = startOffset + parsed.caretOffset.coerceAtLeast(0).coerceAtMost(parsed.text.length)
                        caretModel.moveToOffset(caretOffset)
                    } else {
                        val insertionOffset = context.startOffset
                        doc.insertString(insertionOffset, parsed.text)
                        val caretOffset = insertionOffset + parsed.caretOffset.coerceAtLeast(0).coerceAtMost(parsed.text.length)
                        caretModel.moveToOffset(caretOffset)
                    }
                }
            } else if (textEdit != null && textEdit.isLeft) {
                val edit = textEdit.left
                builder = builder.withInsertHandler { context, _ ->
                    val doc = context.document
                    // 先移除默认插入的 lookupString，避免重复文本
                    if (context.startOffset <= context.tailOffset) {
                        doc.replaceString(context.startOffset, context.tailOffset, "")
                    }
                    // 使用 TextEdit 中定义的范围进行替换
                    val startOffset = toOffset(doc, edit.range.start)
                    val endOffset = toOffset(doc, edit.range.end)
                    doc.replaceString(startOffset, endOffset, edit.newText)
                    context.editor.caretModel.moveToOffset(startOffset + edit.newText.length)
                }
            }
            
            return builder
        }
        
        /**
         * 将 LSP Position 转换为文档偏移量
         */
        private fun toOffset(document: com.intellij.openapi.editor.Document, position: Position): Int {
            val line = position.line.coerceIn(0, document.lineCount - 1)
            val lineStartOffset = document.getLineStartOffset(line)
            val lineEndOffset = document.getLineEndOffset(line)
            val character = position.character
            
            return (lineStartOffset + character).coerceIn(lineStartOffset, lineEndOffset)
        }
        
        /**
         * 根据 CompletionItemKind 获取对应的图标
         */
        private fun getIconForKind(kind: CompletionItemKind?): javax.swing.Icon? {
            if (kind == null) return null
            
            return when (kind) {
                CompletionItemKind.Function -> com.intellij.icons.AllIcons.Nodes.Function
                CompletionItemKind.Variable -> com.intellij.icons.AllIcons.Nodes.Variable
                CompletionItemKind.Class -> com.intellij.icons.AllIcons.Nodes.Class
                CompletionItemKind.Interface -> com.intellij.icons.AllIcons.Nodes.Interface
                CompletionItemKind.Module -> com.intellij.icons.AllIcons.Nodes.Module
                CompletionItemKind.Property -> com.intellij.icons.AllIcons.Nodes.Property
                CompletionItemKind.Keyword -> com.intellij.icons.AllIcons.Nodes.Static
                CompletionItemKind.Method -> com.intellij.icons.AllIcons.Nodes.Method
                CompletionItemKind.Constructor -> com.intellij.icons.AllIcons.Nodes.Class
                CompletionItemKind.Field -> com.intellij.icons.AllIcons.Nodes.Field
                CompletionItemKind.Enum -> com.intellij.icons.AllIcons.Nodes.Enum
                CompletionItemKind.EnumMember -> com.intellij.icons.AllIcons.Nodes.Enum
                CompletionItemKind.Constant -> com.intellij.icons.AllIcons.Nodes.Constant
                CompletionItemKind.Struct -> com.intellij.icons.AllIcons.Nodes.Class
                CompletionItemKind.Event -> com.intellij.icons.AllIcons.Nodes.ErrorIntroduction
                CompletionItemKind.Operator -> com.intellij.icons.AllIcons.Nodes.Function
                CompletionItemKind.Unit -> com.intellij.icons.AllIcons.Nodes.Static
                CompletionItemKind.Value -> com.intellij.icons.AllIcons.Nodes.Variable
                CompletionItemKind.File -> com.intellij.icons.AllIcons.FileTypes.Text
                CompletionItemKind.Snippet -> com.intellij.icons.AllIcons.Actions.Copy
                CompletionItemKind.Color -> com.intellij.icons.AllIcons.Actions.Colors
                CompletionItemKind.Reference -> com.intellij.icons.AllIcons.Nodes.UpLevel
                else -> null
            }
        }

        /**
         * 解析 LSP Snippet 文本，返回可插入的纯文本与光标相对位置
         * 只实现常见占位符：$1、${1}、${1:default}、${1|a,b|}、$0
         * 复杂表达式会被降级为其文字值
         */
        private fun parseLspSnippet(snippet: String): ParsedSnippetResult {
            var textBuilder = StringBuilder()
            var index = 0
            var firstTabstopOffset: Int? = null
            var caretAbsoluteIndexForFirst: Int? = null

            fun recordFirstCaretIfNeeded(currentOutputLen: Int) {
                if (firstTabstopOffset == null) {
                    firstTabstopOffset = currentOutputLen
                }
            }

            while (index < snippet.length) {
                val ch = snippet[index]
                if (ch == '$') {
                    // 尝试匹配 $0 / $1..$9 / ${...}
                    if (index + 1 < snippet.length && snippet[index + 1].isDigit()) {
                        val num = snippet[index + 1]
                        if (num == '0') {
                            // $0 作为最终光标位置，若未设置首个 tabstop，则使用此处
                            recordFirstCaretIfNeeded(textBuilder.length)
                        } else {
                            // $1..$9 将光标定位到第一个 tabstop 出现位置
                            recordFirstCaretIfNeeded(textBuilder.length)
                        }
                        index += 2
                        continue
                    } else if (index + 1 < snippet.length && snippet[index + 1] == '{') {
                        // 处理 ${...}
                        val start = index + 2
                        var i = start
                        var braceDepth = 1
                        while (i < snippet.length && braceDepth > 0) {
                            val c = snippet[i]
                            if (c == '{') braceDepth++
                            if (c == '}') braceDepth--
                            i++
                        }
                        val content = snippet.substring(start, i - 1)
                        // 解析形如 1:default 或 1|a,b|
                        val colonIdx = content.indexOf(':')
                        val pipeStart = content.indexOf('|')
                        val pipeEnd = if (pipeStart >= 0) content.indexOf('|', pipeStart + 1) else -1
                        val defaultText = when {
                            colonIdx >= 0 -> content.substring(colonIdx + 1)
                            pipeStart >= 0 && pipeEnd > pipeStart -> content.substring(pipeStart + 1, pipeEnd).split(',').firstOrNull() ?: ""
                            else -> ""
                        }
                        // 变量编号，用于判断是否首个 tabstop
                        val varNum = content.takeWhile { it.isDigit() }
                        if (varNum.isNotEmpty() && varNum != "0") {
                            recordFirstCaretIfNeeded(textBuilder.length)
                        } else if (varNum == "0") {
                            recordFirstCaretIfNeeded(textBuilder.length)
                        }
                        textBuilder.append(defaultText)
                        index = i
                        continue
                    }
                } else if (ch == '\\' && index + 1 < snippet.length) {
                    // 处理简单转义 \$、\{、\}
                    val next = snippet[index + 1]
                    if (next == '$' || next == '{' || next == '}') {
                        textBuilder.append(next)
                        index += 2
                        continue
                    }
                }
                textBuilder.append(ch)
                index++
            }

            val finalText = textBuilder.toString()
            val caretOffset = (firstTabstopOffset ?: finalText.length)
            return ParsedSnippetResult(finalText, caretOffset)
        }

        /**
         * 解析后的 Snippet 结果
         */
        private data class ParsedSnippetResult(
            val text: String,
            val caretOffset: Int
        )
    }
}